File: Python3/프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.py
def solution(k, room_number):
    result = []
    full = {}
    for room in room_number:
        if room not in full:
            # 빈방이면
            full[room] = room + 1
            result.append(room)
        else:
            # 이미 찬 방이면
            temp = []
            next = room
            while next in full:
                next = full[next]
                temp.append(next)
            
            for t in temp:
                full[t] = next + 1

            result.append(next)
                
    return result

File: Swift/프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.swift
import Foundation

func solution(_ k:Int64, _ room_number:[Int64]) -> [Int64] {
    var graph = [Int64: Int64]()
    var visited = [Int64: Int64]()
    var ans = [Int64]()
    
    for room in room_number {
//        print(room, "--------------------------")
        if visited[room] == nil {
            // 비어있으면 체크인
            visited[room] = room + 1
            ans.append(room)
        } else {
            // 이미 누가 있으면 앞에서부터 빈곳에 채운다.
            var next = room
            var temp = [Int64]()
            while visited[next] != nil {
                temp.append(next)
                next = visited[next]!
            }
            
            visited[next] = next + 1
            for t in temp {
                visited[t] = next + 1
            }
            
            ans.append(next)
        }
    }
    
    return ans
}

